crates/ruff/src/rules/pycodestyle/rules/tab_indentation.rs

```diff
@@ -15,11 +15,53 @@ impl Violation for TabIndentation {
     }
 }
 
+/// `string_lines` is parsed from top to bottom during the tokenization phase, and we know that the
+/// strings aren't overlapping (otherwise there'd only be one string). This function performs a
+/// binary search on `string_lines` to find the string that contains (or starts just before) lineno
+fn find_closest_string<'a>(lineno: usize, string_lines: &'a [Range]) -> Option<&'a Range> {
+    if string_lines.is_empty() {
+        return None;
+    }
+
+    let mut low = 0;
+    let mut high = string_lines.len() - 1;
+
+    while low <= high {
+        let middle = low + (high - low) / 2;
+        if middle == 0 {
+            break;
+        }
+
+        let curr = &string_lines[middle];
+        let start = curr.location.row();
+        let end = curr.end_location.row();
+
+        if start <= lineno && lineno <= end {
+            return Some(curr);
+        } else if start > lineno {
+            high = middle - 1;
+        } else if end < lineno {
+            low = middle + 1;
+        }
+    }
+
+    Some(&string_lines[high])
```

Comment:
You can use rust's built in [`binary-search`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search) method.